core/lib/vm/src/oracles/storage.rs

```diff
@@ -88,12 +90,30 @@ impl<S: WriteStorage, H: HistoryMode> StorageOracle<S, H> {
             || *key == storage_key_for_eth_balance(&BOOTLOADER_ADDRESS)
     }
 
+    fn get_initial_value(&self, storage_key: &StorageKey) -> Option<U256> {
+        self.initial_values.inner().get(storage_key).copied()
+    }
+
+    fn set_initial_value(&mut self, storage_key: &StorageKey, value: U256, timestamp: Timestamp) {
+        if self.get_initial_value(storage_key).is_none() {
+            self.initial_values.apply_historic_record(
+                HashMapHistoryEvent {
+                    key: *storage_key,
+                    value: Some(value),
+                },
+                timestamp,
+            );
+        }
+    }
+
     pub fn read_value(&mut self, mut query: LogQuery) -> LogQuery {
         let key = triplet_to_storage_key(query.shard_id, query.address, query.key);
         let current_value = self.storage.read_from_storage(&key);
 
         query.read_value = current_value;
 
+        self.set_initial_value(&key, current_value, Timestamp(0));
```

Comment:
the timestamp will later be used to rollback the state of the struct, so need to use `query.timestamp`